json-rpc/src/tests.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::runtime::bootstrap;
+use libra_config::utils;
+use reqwest;
+use serde_json;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+use storage_service::mocks::mock_storage_client::MockStorageReadClient;
+
+type JsonMap = HashMap<String, serde_json::Value>;
+
+#[test]
+fn test_json_rpc_protocol() {
+    let address = format!("0.0.0.0:{}", utils::get_available_port());
+    let _runtime = bootstrap(address.parse().unwrap(), Arc::new(MockStorageReadClient));
+    let client = reqwest::blocking::Client::new();
+
+    // check that only root path is accessible
+    let url = format!("http://{}/fake_path", address);
+    let resp = client.get(&url).send().unwrap();
+    assert_eq!(resp.status(), 404);
+
+    // only post method is allowed
+    let url = format!("http://{}", address);
+    let resp = client.get(&url).send().unwrap();
+    assert_eq!(resp.status(), 405);
+
+    // empty payload is not allowed
+    let resp = client.post(&url).send().unwrap();
+    assert_eq!(resp.status(), 400);
+
+    // non json payload
+    let resp = client.post(&url).body("non json").send().unwrap();
+    assert_eq!(resp.status(), 400);
+
+    // invalid version of protocol
+    let request = serde_json::json!({"jsonrpc": "1.0", "method": "add", "params": [1, 2], "id": 1});
+    let resp = client.post(&url).json(&request).send().unwrap();
+    assert_eq!(resp.status(), 200);
+    assert_eq!(fetch_error(resp), -32600);
+
+    // invalid request id
+    let request =
+        serde_json::json!({"jsonrpc": "2.0", "method": "add", "params": [1, 2], "id": true});
+    let resp = client.post(&url).json(&request).send().unwrap();
+    assert_eq!(resp.status(), 200);
+    assert_eq!(fetch_error(resp), -32600);
+
+    // invalid rpc method
+    let request = serde_json::json!({"jsonrpc": "2.0", "method": "add", "params": [1, 2], "id": 1});
+    let resp = client.post(&url).json(&request).send().unwrap();
+    assert_eq!(resp.status(), 200);
+    assert_eq!(fetch_error(resp), -32601);
+
+    // invalid arguments
+    let request = serde_json::json!({"jsonrpc": "2.0", "method": "get_account_state", "params": [1, 2], "id": 1});
+    let resp = client.post(&url).json(&request).send().unwrap();
+    assert_eq!(resp.status(), 200);
+    assert_eq!(fetch_error(resp), -32000);
+}
+
+#[test]
+fn test_get_account_state() {
```

Comment:
Nice!! 
Assuming these hardcoded addresses are in testnet, would it cause problems if testnet needs to be rebooted? Might be worth creating addresses as part of the test

Reply:
It doesn't connect to testnet. It uses mocked storage client(MockStorageReadClient), so as long as it's a valid address, it's going to produce same results